src/ts/ssh/secureStream.ts

```diff
@@ -0,0 +1,235 @@
+	}
+
+	/**
+	 * Initiates the SSH session over the transport stream by exchanging initial messages with the
+	 * remote peer. Waits for the protocol version exchange and key exchange. Additional message
+	 * processing is kicked off as a background promise chain.
+	 * @param cancellation optional cancellation token.
+	 */
+	public async connect(cancellation?: CancellationToken) {
+		if (this.serverCredentials) {
+			const serverSession = <SshServerSession>this.session;
+			serverSession.credentials = this.serverCredentials;
+		}
+
+		await this.session.connect(this.transportStream, cancellation);
+
+		let channel: SshChannel | null = null;
+		if (this.clientCredentials) {
+			let error: SshConnectionError | null = null;
+			const clientSession = <SshClientSession>this.session;
+			if (!(await clientSession.authenticateServer(cancellation))) {
+				error = new SshConnectionError(
+					'Server authentication failed.',
+					SshDisconnectReason.hostKeyNotVerifiable,
+				);
+			}
+
+			if (
+				!error &&
+				!(await clientSession.authenticateClient(this.clientCredentials, cancellation))
+			) {
+				error = new SshConnectionError(
+					'Client authentication failed.',
+					SshDisconnectReason.noMoreAuthMethodsAvailable,
+				);
+			}
+
+			if (error) {
+				await this.session.close(error.reason!, error.message, error);
+				throw error;
+			}
+
+			channel = await this.session.openChannel(cancellation);
+		} else {
+			channel = await this.session.acceptChannel(cancellation);
+		}
+
+		this.stream = this.createStream(channel);
+		// Do not forward the 'readable' event because adding a listener causes a read.
+		this.stream.on('data', (data) => this.emit('data', data));
+		this.stream.on('end', () => this.emit('end'));
+		this.stream.on('close', () => this.emit('close'));
+		this.stream.on('error', () => this.emit('error'));
+		channel.onClosed(() => this.dispose());
+	}
+
+	/**
+	 * Creates a stream instance for a channel. May be overridden to create a `SshStream` subclass.
+	 */
+	protected createStream(channel: SshChannel): SshStream {
+		return new SshStream(channel);
+	}
+
+	public dispose() {
+		if (!this.disposed) {
+			this.disposed = true;
+			this.session.dispose();
+			this.unsubscribe();
+
+			try {
+				if (this.transportStream)
+					this.transportStream.close().catch((e) => {
+						this.trace(
+							TraceLevel.Error,
+							SshTraceEventIds.streamCloseError,
+							`Error closing transport stream: ${e.message}`,
+							e,
+						);
+					});
+			} catch (e) {
+				if (!(e instanceof Error)) throw e;
+				this.trace(
+					TraceLevel.Error,
+					SshTraceEventIds.streamCloseError,
+					`Error closing transport stream: ${e.message}`,
+					e,
+				);
+			}
+		}
+	}
+
+	public async close() {
+		if (!this.disposed) {
+			this.disposed = true;
+
+			await this.session.close(SshDisconnectReason.none, 'SshSession disposed');
+			this.session.dispose();
+			this.unsubscribe();
+
+			await this.transportStream.close();
```

Comment:
What is the difference between `dispose()` and `close()`, should one of the call the other?

Reply:
The main difference is that a session that is closed (but not disposed) due to a connection error may be reconnected, if reconnection is enabled in configuration.

Currently SecureStream doesn't enable reconnection, so the additional dispose call here is redundant. But we might add reconnection support later.